smartsim/_core/entrypoints/dragon.py

```diff
@@ -0,0 +1,178 @@
+Dragon server entrypoint script
+"""
+
+DBPID: t.Optional[int] = None
+
+
+def print_summary(network_interface: str, ip_address: str) -> None:
+    zmq_config = {"interface": network_interface, "address": ip_address}
+
+    with open("dragon_config.log", "w", encoding="utf-8") as dragon_config_log:
+        dragon_config_log.write(
+            textwrap.dedent(f"""\
+                -------- Dragon Configuration --------
+                IPADDRESS: {ip_address}
+                NETWORK: {network_interface}
+                HOSTNAME: {socket.gethostname()}
+                DRAGON_SERVER_CONFIG: {json.dumps(zmq_config)}
+                --------------------------------------
+
+                --------------- Output ---------------
+
+                """),
+        )
+
+
+def run(dragon_head_address: str) -> None:
+    global SHUTDOWN_INITIATED  # pylint: disable=global-statement
+    print(f"Opening socket {dragon_head_address}")
+    dragon_head_socket = context.socket(zmq.REP)
+    dragon_head_socket.bind(dragon_head_address)
+    dragon_backend = DragonBackend()
+
+    while not SHUTDOWN_INITIATED:
+        print(f"Listening to {dragon_head_address}")
+        req = dragon_head_socket.recv_json()
+        print(f"Received request: {req}")
+        drg_req = request_serializer.deserialize_from_json(str(req))
+        resp = dragon_backend.process_request(drg_req)
+        print(f"Sending response {resp}", flush=True)
+        dragon_head_socket.send_json(response_serializer.serialize_to_json(resp))
+        if isinstance(resp, DragonShutdownResponse):
+            SHUTDOWN_INITIATED = True
+
+
+def main(args: argparse.Namespace) -> int:
+    interface, address = get_best_interface_and_address()
+    if not interface:
+        raise ValueError("Net interface could not be determined")
+    dragon_head_address = f"tcp://{address}"
+
+    if args.launching_address:
+        if str(args.launching_address).split(":", maxsplit=1)[0] == dragon_head_address:
+            address = "localhost"
+            dragon_head_address = "tcp://localhost:5555"
+        else:
+            dragon_head_address += ":5555"
+
+        launcher_socket = context.socket(zmq.REQ)
+        launcher_socket.connect(args.launching_address)
+
+        response = (
+            _helpers.start_with(DragonBootstrapRequest(address=dragon_head_address))
+            .then(request_serializer.serialize_to_json)
+            .then(launcher_socket.send_json)
+            .then(lambda _: launcher_socket.recv_json())
+            .then(str)
+            .then(response_serializer.deserialize_from_json)
+            .get_result()
+        )
+
+        if not isinstance(response, DragonBootstrapResponse):
+            raise ValueError(
+                "Could not receive connection confirmation from launcher. Aborting."
+            )
+
+    print_summary(interface, dragon_head_address)
+
+    run(dragon_head_address=dragon_head_address)
+
+    print("Shutting down! Bye bye!")
+    return 0
```

Comment:
We don't catch any exceptions in here. Can we add something so we ensure errors are logged?

Reply:
Added a generic `except Exception` and returned a `os.EX_SOFTWARE` code.